### ISSUE TITLE
winrt: open device ReadWrite fails

### ISSUE THREAD

**Issue by zardini123** [sidit77/async-hid#7]
When trying the example `read_write` on some devices I own, every time the process does not exit successfully with the error:

```
Error: HidError: Custom("Failed to open \\\\?\\HID#VID_203A&PID_FFFC&MI_02#7&34ad9e6b&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}")
        at C:\...\async-hid\src\backend\winrt\mod.rs:124:28
```

I tried opening with ReadWrite on my mouse, keyboard, PS4, and PS5 controllers, and all failed.  I have integrated `async-hid` into a personal project and `async-hid` has an identical problem there.

The only AccessMode that works is `Read`.  I can successfully acquire read reports with all my devices listed.

Opening devices for read/write on Windows does not happen when using [`hidapi`](https://github.com/ruabmbua/hidapi-rs).

Any help would be greatly appreciated.  Thanks!

**Comment by sidit77** [comment]
Some of these are unfortunatly not supported:

> The Windows.Devices.HumanInterfaceDevice API supports most HID devices. However, it blocks the top-level application collection represented by the following usage pages, to prevent conflict with other Windows APIs and OS behavior:
>
>    HID_USAGE_PAGE_UNDEFINED
>    HID_USAGE_PAGE_GENERIC
>    HID_USAGE_GENERIC_KEYBOARD
>    HID_USAGE_GENERIC_KEYPAD
>    HID_USAGE_GENERIC_SYSTEM_CTL
>    HID_USAGE_PAGE_KEYBOARD
>    HID_USAGE_PAGE_CONSUMER
>    HID_USAGE_PAGE_DIGITIZER
>    HID_USAGE_PAGE_SENSOR
>    HID_USAGE_PAGE_BARCODE_SCANNER
>    HID_USAGE_PAGE_WEIGHING_DEVICE
>    HID_USAGE_PAGE_MAGNETIC_STRIPE_READER
>    HID_USAGE_PAGE_TELEPHONY

This is the relevent section of hidapi:
```rust
let handle = open_device(&device_path, true)
        // System devices, such as keyboards and mice, cannot be opened in
        // read-write mode, because the system takes exclusive control over
        // them.  This is to prevent keyloggers.  However, feature reports
        // can still be sent and received.  Retry opening the device, but
        // without read/write access.
        .or_else(|_| open_device(&device_path, false))?;
```

I don't think it's possible to replicate with winrt as `FileAccessMode` can only be `Read` or `ReadWrite`

**Comment by zardini123** [comment]
Thanks for your response!  I have some follow-up questions.

> Some of these are unfortunatly not supported:

This still begs the question why my gamepads (PS4, PS5) aren't opening.  Is there something I am missing when calling your API?

> This is the relevent section of hidapi:

Where can I find the source of this section?  I am having troubles finding it.

> I don't think it's possible to replicate with winrt as `FileAccessMode` can only be `Read` or `ReadWrite`

What would be the ultimate solution be then for these devices?  Does `hidapi` use a different windows API that is different to acquire access?

**Comment by sidit77** [comment]
Just for clarification: Are you trying to open the gamepad in read or read/write mode?

> Where can I find the source of this section? I am having troubles finding it.

Original C code: https://github.com/libusb/hidapi/blob/d0732cda906ad07b7e1ef93f1919035643620435/windows/hid.c#L991
My Rust port: https://github.com/ruabmbua/hidapi-rs/blob/2f37587cc7d7d56f2cdf3c948abffac860d63b5c/src/windows_native/mod.rs#L398
(`hidapi` could use either depending on if you activated the `windows_native` feature)

> Does hidapi use a different windows API that is different to acquire access?

Yes. What hidapi essentially uses the win32 way of opening devices, while this library uses the higher level winrt classes (virtually no `unsafe`, automatic async, at lot less code). The reason the backend is called `winrt` instead of `windows` is because I thought that it might become nessecary to add a second Windows backend (`win32`).

**Comment by zardini123** [comment]
> Just for clarification: Are you trying to open the gamepad in read or read/write mode?

Read/write mode :)

> (`hidapi` could use either depending on if you activated the `windows_native` feature)

Thanks for the links!  From my brief scan of the `windows-native` section of `async-hid`, it seems the implementation is similar to the `winrt` implementation of this library, but synchronized.  So is the `windows_native` part of `hidapi` using `winrt`, and the C implementation is using `win32`?

> Yes. What hidapi essentially uses the win32 way of opening devices, while this library uses the higher level winrt classes (virtually no `unsafe`, automatic async, at lot less code). The reason the backend is called `winrt` instead of `windows` is because I thought that it might become nessecary to add a second Windows backend (`win32`).

Oh, that is really interesting!  I appreciate the forward-thinking on the backend organization.

My thought is until the `win32` backend is implemented, I think it would be useful to notify users why those devices are not opening is because it is inherit to the `winrt` backend.  Would it be possible to check if the `HID_USAGE` for the device attempting to be opened is any of those usage types you mentioned, and then throw a more crafted error for this scenario?

**Comment by sidit77** [comment]
> So is the windows_native part of hidapi using winrt, and the C implementation is using win32?

No both of them are using win32. The windows-native section is more or less a one-to-one port of the C implementation.


A lot of these restrictions are not really inherent to winrt but Windows as a whole. Other operating systems have other restrictions. On Linux for example you can't open any device unless you're either running your program as root or installed the appropriate udev rules. On MacOS you'll get a permission popup for the user, that can be denied and probably more. With WASM/WebHID (not yet in main) you'll have to manually ask for a similar permission popup before you can even enumerate devices and enumerate will only find devices that the user granted access to.

**Comment by Tederean** [comment]
I also have the problem that I cannot open a device writeable with the WinRT API, although I think the device should be compatible. UsagePage is 0xFF00, UsageId: 0x0001. According to the Microsoft documentation, 0xFF00/Vendor Specific is not on the blacklist.

The USB device works writeable under the Win32 API, under Linux, macOS, Android and WebHID. But this line always returns zero...

[https://github.com/sidit77/async-hid/blob/7418bdda19729c5aa3ddca0996beab5bd89c13f7/src/backend/winrt/mod.rs#L57C5-L60C92
](url)

To avoid misunderstandings, can a normal Rust programm with the target x86_64-pc-windows-msvc interact with the WinRT API at all? The Microsoft documentation states that the application must specify in the manifest which UsagePage the application wants to control in order for everything to work. Do I have to do the same?

**Comment by sidit77** [comment]
> To avoid misunderstandings, can a normal Rust programm with the target x86_64-pc-windows-msvc interact with the WinRT API at all?

Yes (at least on my PC). I wrote this library for [my replacement for the steelseries engine](https://github.com/sidit77/headset-controller) and it worked without any issues.

When you package a Windows binary you can set it's trust level to "appContainer" or "mediumIL".  mediumIL is only supported on desktop enviroments and is essentially what you get with a normal, unpackaged, Win32 exe, appContainer is what requires you to declare your capabilites in the manifest. However the situation is pretty poorly documented and I also encountered issues with other apis that should work but don't for some reason. 

I think maybe it's better to rewrite the Windows backend with win32 regardless of how nice the WinRT apis are since they seem to have these weird issues. I'll have to see when I find the time and motivation to work on this.

**Comment by sidit77** [comment]
I created a WIP [branch](https://github.com/sidit77/async-hid/tree/move-to-win32) for the win32 port. Neither read or write are currently async but they should work. It would be nice if you could confirm if they work for you as well, @Tederean .

**Comment by Tederean** [comment]
I would like to avoid us getting in each other's way. Please let me know how you would like to proceed and how I can help.

**Comment by sidit77** [comment]
Great to hear that it's now working.

> How about if uwp is selected as target, the uwp API is used and otherwise the Win32 API?

I would just drop the uwp target, tbh. I would've been nice to have, but I don't think it justifies an additional backend.

From my perspective, there are two areas that still need some work:

1. Device enumeration. Right now, I'm still using the uwp APIs for that, I would eventually move these over to win32 as well.

2. Async. To implement async we need a way to wait on Windows events handles. I've looked into that for both `tokio` and `async-io` and only `async-io` has primitives that help with this. I've looked a bit into the implementation and it seems like they just delegate the task to the Windows threadpool using [RegisterWaitForSingleObject](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-registerwaitforsingleobject). To maintain compatibility with both Tokio and async-io I would just use this (or another threadpool function) directly. The big issue here is to make sure that the read function remains cancellation save. I wanted to do some experiments in this direction this weekend if I find the time.

**Comment by Tederean** [comment]
I think it would be a great pity to simply dispose of a good and clear backend. Yes, it has weaknesses under a normal Win32 application, but this backend is designed for UWP and there it is the only way to interact with HID devices. This backend is great for UWP.

There is a cfg option to recognise UWP as a target. I would recommend removing the serial number extension for UWP via the Win32 API and leaving it as it is. Due to the clear source code and API, the maintenance effort should be very low.

I would be very sorry to deprive a Rust library of this existing possibility, as this library is intended to enable HID interaction on many platforms.

I would offer myself for a refractoring. I could still do that today. I have seen that e.g. some functions can be replaced by safe Rust std alternatives. E.g. opening the file:

```
        let result = OpenOptions::new()
            .read(access_mode.readable())
            .write(access_mode.writeable())
            .create(false)
            .share_mode(FILE_SHARE_READ.0 | FILE_SHARE_WRITE.0)
            .custom_flags(FILE_FLAG_OVERLAPPED.0)
            .open(path);

        result.unwarp().as_raw_handle()
```

I could also submit a division or a proposal for how Win32 and UWP could be separated.

**Comment by sidit77** [comment]
True, keeping the UWP backend around is probably not too much trouble, other than the need for extension traits.

I think we should simply introduce two features `win32` and `uwp` and add one of them as default feature. Eanbling both should probably be a compile time error.

I worked a bit on the async read implementation and it should be in a pretty good state now. However I did notice that `pub async fn read_input_report(&self, buf: &mut [u8])` should really take `&mut self` instead as two thread/tasks reading from the same device instance at the same time just makes no sense. Currently most backends deal with this issue by just giving each new packet to a random reader, which doesn't sound useful to me. The same should honestly apply to the write functions. Worst case the user can just wrap the device into a mutex themselves.

Regarding the `OpenOptions`: I would prefer to keep using `OpenFile` directly as it keeps the used APIs more consistent. I'd guess that open options internally just calls `OpenFile` anyways.

**Comment by Tederean** [comment]
> True, keeping the UWP backend around is probably not too much trouble, other than the need for extension traits.
> 
> I think we should simply introduce two features win32 and uwp and add one of them as default feature. Eanbling both should probably be a compile time error.

I think that sounds good regarding UWP. 

> However I did notice that pub async fn read_input_report(&self, buf: &mut [u8]) should really take &mut self instead as two thread/tasks reading from the same device instance at the same time just makes no sense. Currently most backends deal with this issue by just giving each new packet to a random reader, which doesn't sound useful to me. The same should honestly apply to the write functions. Worst case the user can just wrap the device into a mutex themselves.

The problem is that you can no longer wait for a report to be read and send a report at the same time if we make the functions mutable. This is exactly the use case I need.

Yes, reading or writing two tasks/threads at the same time makes no sense, but you could manage that with a mutex.

I am developing a device that has many different reports that do not come back in the same order as they were sent. There are reports that take several seconds to be responded by the device. The device interacts with sensors, some take longer than others to respond, which are then forwarded directly to the host. If I now wait for a long command, I can no longer send anything. With hidapi I could still poll, but that's exactly what async is supposed to do better.

> Regarding the OpenOptions: I would prefer to keep using OpenFile directly as it keeps the used APIs more consistent. I'd guess that open options internally just calls OpenFile anyways.

Sure.

**Comment by sidit77** [comment]
That's a good point. In this case I'll just leave the [internal Mutex implementation](https://github.com/sidit77/async-hid/blob/move-to-win32/src%2Fbackend%2Fwinrt%2Fmod.rs#L148). We should probably add some documentation to clarify that having more than one "actor" per operation is unsupported and platform-specific behavior.

**Comment by Tederean** [comment]
We could split Reader and Writer to seperate Structs while using &mut self instead of interiour mutability.

Embassy uses this approach on device side for HID.

Later I have more time to explain this in detail…

**Comment by sidit77** [comment]
I know what you're talking about. Tokio does to same for TcpConnections. I generally like the idea, but it's a bigger refactor that applies to all backends. For this reason I would push this back until the win32 is merged into main.

**Comment by Tederean** [comment]
I have just had a look at the win32 source code, excellent work.

> I know what you're talking about. Tokio does to same for TcpConnections. I generally like the idea, but it's a bigger refactor that applies to all backends. For this reason I would push this back until the win32 is merged into main.
Ok.

I offer to do the refactoring work for Reader & Writer when you are done with the Windows backend. Get back to me if you want.

**Comment by sidit77** [comment]
From my end, all that's left to do is implement the backend selection. I'll try to finish this over the next weekend.

**Comment by Tederean** [comment]
Currently Send trait is no longer satisfied because of Box usage in IoBuffer.

```
pub struct IoBuffer<T> {
    device: Arc<Device>,
    buffer: Box<[u8]>,
    overlapped: Box<Overlapped>,
    pending: bool,
    _marker: PhantomData<T>,
}
```

Same problem as with macos backend back then. We can replace it with Arc and optional Mutex or we do it the raw way like in the macos backend currently. I prefer the second option.

I have taken the liberty of implementing a backend selection and created a PR for this.

**Comment by sidit77** [comment]
Are you sure? I added a [Send + Sync test](https://github.com/sidit77/async-hid/blob/1e2cd87eb5e33bba2573c11f825600beb6e31c15/src/lib.rs#L162) to lib.rs and it runs without any issues. I also just modifed one of the examples to move the device between threads and it also compiles:
```rust
let device = DeviceInfo::enumerate()
        .await?
        .find(|info: &DeviceInfo| info.matches(0xFFC0, 0x1, 0x1038, 0x2206))
        .await
        .inspect(|info| {
            println!(
                "{}: 0x{:X} 0x{:X} 0x{:X} 0x{:X} {:?}",
                info.name,
                info.usage_page,
                info.usage_id,
                info.vendor_id,
                info.product_id,
                info.id
            );
        })
        .expect("Could not find device")
        .open(AccessMode::ReadWrite)
        .await?;
    std::thread::spawn(|| pollster::block_on(async move {
        device.write_output_report(&[0x0, 0xb0]).await?;
        let mut buffer = [0u8; 8];
        let size = device.read_input_report(&mut buffer).await?;
        println!("{:?}", &buffer[..size]);
        Ok::<(), HidError>(())
    })).join().unwrap()?;
```

The docs [1](https://doc.rust-lang.org/std/boxed/struct.Box.html#impl-Send-for-Box%3CT,+A%3E) [2](https://doc.rust-lang.org/std/primitive.slice.html#impl-Send-for-%5BT%5D) also seem to agree. 

Is it possible that you're not on the latest commit? Because the Send + Sync test also failed for me until I implemented `Send` for `Overlapped`.

**Comment by Tederean** [comment]
Same error.

**Comment by sidit77** [comment]
Sorry, I missed the second part of the message. I don't think you can make this future `Send` as it is because of how the std `Mutex` is implemented. If you really need to make this future `Send` then we would have to implement a custom mutex with a guard that is `Send` (or use a library).

I agree that the Future returned by `read` should be `Send` because it won't work on multithreaded runtimes otherwise. I believe doing the reader/writer split would allow us to completely remove the mutex as both could simply use `&mut self`.

**Comment by sidit77** [comment]
I pushed a new commit that implements the mentioned custom mutex and fixes the `Send` issue

**Comment by Tederean** [comment]
It now compiles and works for me. However, I can't tell at the moment whether we have created undefined behaviour under certain conditions in this backend.

From my side, we can merge into main, when you accepted by PR. After or during the refactoring, I think it makes more sense to take a closer look at possible UB.

**Comment by sidit77** [comment]
Yeah, this backend is kinda scary. I did my best to follow MSDN and program defensively but it's very easy to overlook something.

Also WebHID was also on my todo list. There's even an unifinished branch for WebHID in this repo. So it's great to see some progress there.
